File: src/api/subsonic/methods/getArtists.ts
import * as v from 'valibot';

import {
  handleJsonResponse,
  makeSubsonicRequest,
  parseJsonResponse,
} from '../makeRequest';
import { BaseArtist } from '../types/baseArtist';

export function subsonicGetArtists() {
  return makeSubsonicRequest({ method: 'rest/getArtists' })
    .flatMap(response =>
      parseJsonResponse(response, {
        artists: v.object({
          ignoredArticles: v.string(),
          index: v.array(
            v.object({ artist: v.array(BaseArtist), name: v.string() }),
          ),
          lastModified: v.optional(v.number()),
        }),
      }),
    )
    .flatMap(json => handleJsonResponse(json['subsonic-response']))
    .map(result => result.artists);
}

File: src/api/subsonic/getArtists.ml
open Fx
open Monad_syntax
open Utils

let make () =
  let* { response = { status = `ok; artists } } =
    let params = Js.Dict.empty () in
    makeRequest { _method = "rest/getArtists"; params } Artists.of_json
  in
  Ok artists
